ledger-core/virtual/handlers/vcachedmemoryrequest.go

```diff
@@ -6,35 +6,131 @@
+	memoryCache   memoryCacheAdapter.MemoryCache
+	pulseSlot     *conveyor.PulseSlot
+
+	object   descriptor.Object
+	response *payload.VCachedMemoryResponse
 }
 
 /* -------- Declaration ------------- */
+
 var dSMVCachedMemoryRequestInstance smachine.StateMachineDeclaration = &dSMVCachedMemoryRequest{}
 
 type dSMVCachedMemoryRequest struct {
 	smachine.StateMachineDeclTemplate
 }
 
-func (*dSMVCachedMemoryRequest) InjectDependencies(_ smachine.StateMachine, _ smachine.SlotLink, _ injector.DependencyInjector) {
+func (*dSMVCachedMemoryRequest) InjectDependencies(sm smachine.StateMachine, _ smachine.SlotLink, injector injector.DependencyInjector) {
+	s := sm.(*SMVCachedMemoryRequest)
+
+	injector.MustInject(&s.pulseSlot)
+	injector.MustInject(&s.messageSender)
+	injector.MustInject(&s.memoryCache)
 }
+
 func (*dSMVCachedMemoryRequest) GetInitStateFor(sm smachine.StateMachine) smachine.InitFunc {
 	s := sm.(*SMVCachedMemoryRequest)
 	return s.Init
 }
 
 /* -------- Instance ------------- */
+
+func (*SMVCachedMemoryRequest) InjectDependencies(sm smachine.StateMachine, _ smachine.SlotLink, injector injector.DependencyInjector) {
+	s := sm.(*SMVCachedMemoryRequest)
+
+	injector.MustInject(&s.pulseSlot)
+	injector.MustInject(&s.memoryCache)
+	injector.MustInject(&s.messageSender)
+}
+
 func (s *SMVCachedMemoryRequest) GetStateMachineDeclaration() smachine.StateMachineDeclaration {
 	return dSMVCachedMemoryRequestInstance
 }
+
+func (s *SMVCachedMemoryRequest) migrationDefault(ctx smachine.MigrationContext) smachine.StateUpdate {
+	ctx.Log().Trace("stop processing VCachedMemoryRequest since pulse was changed")
+	return ctx.Stop()
+}
+
 func (s *SMVCachedMemoryRequest) Init(ctx smachine.InitializationContext) smachine.StateUpdate {
+	if s.pulseSlot.State() != conveyor.Present {
+		ctx.Log().Trace("stop processing VCachedMemoryRequest since we are not in present pulse")
+		return ctx.Stop()
+	}
+	ctx.SetDefaultMigration(s.migrationDefault)
+	return ctx.Jump(s.getMemory)
+}
+
+func (s *SMVCachedMemoryRequest) getMemory(ctx smachine.ExecutionContext) smachine.StateUpdate {
+	s.memoryCache.PrepareAsync(ctx, func(ctx context.Context, svc memorycache.Service) smachine.AsyncResultFunc {
+		obj, err := svc.Get(ctx, s.Payload.Object)
+		return func(ctx smachine.AsyncResultContext) {
+			s.object = obj
+			if err != nil {
+				ctx.Log().Error("failed to get memory", err)
+			}
+		}
+	}).Start()
+	return ctx.Jump(s.waitResult)
+}
+
+func (s *SMVCachedMemoryRequest) waitResult(ctx smachine.ExecutionContext) smachine.StateUpdate {
+	if s.object == nil {
+		return ctx.Sleep().ThenRepeat()
+	}
+	return ctx.Jump(s.buildResult)
+}
+
+func (s *SMVCachedMemoryRequest) buildResult(ctx smachine.ExecutionContext) smachine.StateUpdate {
+	if s.object.HeadRef().IsEmpty() {
+		s.response = &payload.VCachedMemoryResponse{
+			Object:     s.Payload.Object,
+			StateID:    s.Payload.StateID,
+			CallStatus: payload.CachedMemoryStateMissing,
```

Comment:
I think this should be unknown.

Reply:
What descriptor will be considered as Missing? with empty memory?

Reply:
not with empty memory. Read documentation: unknown - "I don't have, but not sure about others", missing - "not existent in the whole system".

Reply:
fixed